### DISCUSSION TITLE
Non-reliable saving

### DISCUSSION THREAD

**Discussion by liadJB** [bonsai-rx/bonsai#1758] [Q&A]
Hi,

I am attaching a script we are using. We are having problems whereby all the files are not consistently saved. Always on the first run of the protocol not everything is saved. The chances of saving are higher on the second run, but still sometimes not.
The thing that's not always saving is the arudino input:
![image](https://github.com/bonsai-rx/bonsai/assets/48135253/072c80ac-2bc1-484d-8843-6a3afb2d4f94)

Any idea why it might happen?

[Gratings.zip](https://github.com/bonsai-rx/bonsai/files/15108353/Gratings.zip)

Thanks

**Comment by bruno-f-cruz** [comment]
Can you provide a little more details? Is it the case that data is missing? The full file is missing? part of the last row is not showing? Do you have a sense how many samples are missing? How did you arrive to this conclusion?

Additionally, can you provide a minimal example that reproduces the bug without having to run the full workflow? Thanks

**Comment by liadJB** [comment]
Thanks. I will try to see if I can reproduce it without the rest. In the
meantime. The file is created with 0kb. So it seema like there is some sort
of race condition that sometimes prevents it being written.

**Reply by bruno-f-cruz** [reply]
Are you sure you have events taking place downstream from your condition node? The CSV will be open at subscription time, NOT when the first sample arrives.

**Comment by liadJB** [comment]
Yep. Nothing changes between the firat run and the second. Stimuli are
oresented and I can view everything.

**Reply by bruno-f-cruz** [reply]
So if you inspect the output of the condition node while bonsai is running you see values in both "runs"? (What do you mean by runs btw?)

**Comment by liadJB** [comment]
Yes. One can see values in all nodes. By run I mean pressing the start
button and stopping.
If one then clises Bonsai one needs to 'run' a script first again.

**Comment by Awlala** [comment]
Thank you for raising this question. I have the same problem. I have not yet been able to solve it and look forward to get some hints here. I was also wondering if you had any error messages raised in the log. It seems that it may provide more information about the problem and when my .csv files save properly, the error message does not show up.

**Reply by bruno-f-cruz** [reply]
Can you have an example that minimally reproduced this behavior?

**Reply by Awlala** [reply]
Sorry I may not fully understand the request correctly, still very new to this. I assume you are referring to an example of the error message in log that appear when the files are not saved properly? If so, here is the message:
2024-04-26T17:01:02.196ZE [1092:NonCelloThread] registry_win.h:55:GetProtoFromRegistryValue Opening registry key Software\Google\DriveFS\Share failed with 0x2
2024-04-26T17:01:02.196ZE [12740:NonCelloThread] registry_win.h:55:GetProtoFromRegistryValue Opening registry key Software\Google\DriveFS\Share failed with 0x2
2024-04-26T17:01:11.211ZE [1092:NonCelloThread] registry_win.h:55:GetProtoFromRegistryValue Opening registry key Software\Google\DriveFS\Share failed with 0x2

**Reply by bruno-f-cruz** [reply]
I was referring to the bonsai workflow you are using to acquire data with. Can you provide a simple example that I can run on my machine to reproduce the weird behavior you are referring to

**Reply by Awlala** [reply]
Oh I see. I will try to get a simpler example. Thank you for the reply.

**Reply by liadJB** [reply]
Hi, please find attached a minimal example.
On the very first run with the same Bonsai instance. Neither the log nor the arduino data (relying on CSV writing) are saving.
If you run it again, in the same bonsai instance, all is saved. 
If you close Bonsai, you'd need to start the process of a 'dummy' run 
and then an actual run again. 
[MinimalExampel.zip](https://github.com/bonsai-rx/bonsai/files/15160240/MinimalExampel.zip)

**Reply by bruno-f-cruz** [reply]
I am sorry but I dont have an Arduino, Spinnaker camera nor a national instruments board to be able to reproduce you workflow, can you reproduce the behavior without using these devices?

Otherwise, I would probably just start logging data at multiple places of the pipeline and check where data is being lost. For instance, what happens when you compare the output of these 4 CSVs?

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/42007030-4383-4353-a050-0e51d9c51c78)

**Reply by liadJB** [reply]
So I did some tests.
It seems when I delay subscription to the serial port reader by 10ms it consistently saves everything on the very first run.
So it does suggest a race condition between opening the port and reading from it that bonsai does not deal with. 
What I do not understand why without this delay, it works only from the second run onwards.

**Reply by liadJB** [reply]
Hi,

I ran the script as you suggested. When the data isn't saved, none of the csv files have any data in them.
Also if I have a viewer on the CreateSerialPort node nothing is shown.
It seems like the problem is that sometimes CreateSerialPort just doesn't work. (DtrEnabled is True).
Is there a smart way to use While to try and start the port until it actually starts?
If so, I would really appreciate an example snippet.

Liad

**Reply by bruno-f-cruz** [reply]
I would encourage you to find out why the Device is not connecting correctly. It may be a different problem altogether. You can try to go around this in several ways and keep trying to reconnect. One way could be using `TimeOut` and `Retry`:
(I have not tested this as I don't have access to a serial device right now)

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/16d2fecc-0085-41ea-acfd-b9362f1c4df2)

<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.8.1"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns:port="clr-namespace:Bonsai.IO.Ports;assembly=Bonsai.System"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="rx:Defer">
        <Name>SerialObject</Name>
        <Workflow>
          <Nodes>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="port:CreateSerialPort">
                <port:BaudRate>9600</port:BaudRate>
                <port:NewLine>\r\n</port:NewLine>
                <port:Parity>None</port:Parity>
                <port:ParityReplace>63</port:ParityReplace>
                <port:DataBits>8</port:DataBits>
                <port:StopBits>One</port:StopBits>
                <port:Handshake>None</port:Handshake>
                <port:DiscardNull>false</port:DiscardNull>
                <port:DtrEnable>false</port:DtrEnable>
                <port:RtsEnable>false</port:RtsEnable>
                <port:ReadBufferSize>4096</port:ReadBufferSize>
                <port:WriteBufferSize>2048</port:WriteBufferSize>
                <port:ReceivedBytesThreshold>1</port:ReceivedBytesThreshold>
              </Combinator>
            </Expression>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="port:SerialReadLine" />
            </Expression>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="rx:Timeout">
                <rx:DueTime>PT0S</rx:DueTime>
              </Combinator>
            </Expression>
            <Expression xsi:type="WorkflowOutput" />
          </Nodes>
          <Edges>
            <Edge From="1" To="2" Label="Source1" />
            <Edge From="2" To="3" Label="Source1" />
          </Edges>
        </Workflow>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Retry" />
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

```

</details>

**Reply by liadJB** [reply]
The only difference is that we are using the new Bonsai version. All the rest, is exactly as before. 
If anyone knows the exact changes to the Serial handling code, it would be good to check.
In the meantime, thanks for the work around I will try and use that.

**Reply by bruno-f-cruz** [reply]
Do you have access to the previous environment? I would be very interested to know if you could reliability reproduce this bug in the new version and not the previous one.

**Reply by liadJB** [reply]
Unfortunately not. I updated it (unsuccessfully) and needed a tailor made version from you.
And it seemed impossible to fully downgrade, because of package conflicts etc.

**Reply by liadJB** [reply]
Thanks for the suggestion. All it does it throw a timeout exception though. 
Would you have any other suggestion? 
Maybe a similar idea, but with a toggle that says that first data was received and wait on this toggle and as long as it is not received try to connect?

**Reply by bruno-f-cruz** [reply]
I am not sure I follow. What I sent shouldn't throw an exception. Retry should keep trying to subscribe to the serial connection. Can you share the workflow you are using?

**Reply by liadJB** [reply]
Here is the example.
In the cases when the timeout is triggered there's a timeout exception, otherwise it's working as usual (in the good times, when the connection is actually established)

[Example.zip](https://github.com/bonsai-rx/bonsai/files/15399934/Example.zip)

**Reply by glopesdev** [reply]
@liadJB there is a chance this could be a bug in the operator. If the error appears in the status bar at the bottom of the Bonsai editor, can you right-click the text and copy the full stack trace here?

**Reply by liadJB** [reply]
I also tried to implement it as Bruno suggested.

If I take the include the circled parts in a Defer group, there is an error saying the Arduino subject does not exist in the context of the part circled in blue. 

![image](https://github.com/bonsai-rx/bonsai/assets/48135253/62e73d2d-3b35-4727-a526-84d6a238a3dc)

Would there be any way to run a quick loop at the start of the run to verify that the serial port is working and somehow then continue ?

**Reply by liadJB** [reply]
Hi,

I am re-flagging it, since we updated the bonsai script in another of our rigs (without changing anything else - so the hardware and Arduino remained the same), and we are now having the same problem. So we can be fairly certain that the problem lies in the updated operators.
Since it is asynchronous it's hard to say what is happening, but maybe a comparison of what change between version could help understand what changed between the versions?

**Comment by liadJB** [comment]
Hi,

There is just an error box saying that a Timeout exception was thrown.
